### PR TITLE
[WIP] Parse Async Generators

### PR DESCRIPTION
**Pull request by mikeyhew**
| Q                 | A
| ----------------- | ---
| Bug fix?          | no
| Breaking change?  | no
| New feature?      | yes
| Deprecations?     | no
| Spec compliancy?  | yes (stage3)
| Tests added and they pass? | yes
| Fixed tickets     | lightscript/lightscript#15
| License           | MIT

So far this supports parsing `-/*>` and `=/*>`. There are two new tests, both of which only test parsing, and only for the named function syntax.

I didn't wrap the new parsing code in an `if` statement because I wasn't sure which plugin name to use. [This file](https://github.com/babel/babel/blob/master/packages/babel-plugin-syntax-async-generators/src/index.js) would suggest that it is `asyncGenerators`, is that correct?

### CODE REVIEW

File: src/plugins/lightscript.js [hunk 1]
```diff
@@ -302,6 +302,10 @@ pp.parseArrowType = function (node) {
     case "=*>": case "-*>":
       node.generator = true;
       break;
+    case "=/*>": case "-/*>":
```
Comment (rattrayalex) [inline src/plugins/lightscript.js]: Sorry, could we do `=*/>` and `-*/>` instead? I prefer it visually and it's less likely to trip up syntax highlighters who think it's a comment. Unless there's a reason you prefer this style?

File: src/tokenizer/index.js [hunk 2]
```diff
@@ -400,6 +400,14 @@ export default class Tokenizer {
         return this.finishToken(tt.arrow, "-*>");
       }
 
+      if (this.hasPlugin("asyncGenerators")) {
```
Comment (rattrayalex) [inline src/tokenizer/index.js:404]: nice, this looks good!

File: test/fixtures/lightscript/named-arrow-functions/skinny-async-gen/actual.js [hunk 3]
```diff
@@ -0,0 +1,2 @@
+fn() -/*>
+  yield await x
```
Comment (rattrayalex) [inline test/fixtures/lightscript/named-arrow-functions/skinny-async-gen/actual.js]: these look good – let's also add the following tests:

```
f() =*/>
  yield <- g()
```


```
f() =*/>
  yield <!- g()
```

```
f() =*/>
  x = yield <- g()
```

These tests will also probably need to be copied into the plugin's test suite

Comment (rattrayalex) [reply test/fixtures/lightscript/named-arrow-functions/skinny-async-gen/actual.js]: We should also probably test that this raises a SyntaxError:

```
f() =*/>
  <- yield g()
```

If it doesn't do that by default, let's make it a friendly one 😄

**Review by rattrayalex — COMMENTED** [review]
Great work! One change request and a few additional tests that I imagine should be straightforward. 

I can't remember, does anything else need to happen for this?

File: src/plugins/lightscript.js (hunk 1 above)
Comment (mikeyhew) [reply src/plugins/lightscript.js]: `=/*>` is what came naturally to me, but sure, I can change it.

Comment (rattrayalex) [reply src/plugins/lightscript.js]: Thanks!

**Comment by rattrayalex** [comment]
Hey @mikeyhew are you still interested in taking this over the finish line? I can help if not!

**Comment by rattrayalex** [comment]
(note that there are likely to be significant merge conflicts, which I can help you with as they're fairly straightforward)

**Comment by mikeyhew** [comment]
@rattrayalex Sorry, I got sucked into Rust. (The programming language, not the game. 🙂 ) I'll see what I can get done tonight, and then after that I'll let you take over since I have school starting up again.

**Comment by mikeyhew** [comment]
@rattrayalex I'll leave it for you to finish. I think I got the conflict sorted out. There's still those tests to add that you suggested; plus the tests that I wrote aren't passing anymore, and I didn't really understand the error message - "Did not expect a property 'delegate' (expression/0/body/body/0/body/program)".

I wanted to point out about the one negative test you suggested:

```
f() =*/>
  <- yield g()
```

You're allowed to have `yield` expressions inside `await` expressions - it just awaits the return value of the `yield` expression. But `await yield g()` seems to be a syntax error, I guess because `await` has higher precedence? Anyway, `await (yield g())` is correct and parses fine.

**Comment by rattrayalex** [comment]
Awesome! I can take it from here then. The delegate thing is probably from the upgrade - removing the expected.json file and then re-running the test suite is probably the easiest way to fix it.

**Comment by rattrayalex** [comment]
@mikeyhew finally merged, sorry this took me so long to get to! Thanks for the contribution!

**Comment by mikeyhew** [comment]
@rattrayalex no worries, thanks for finishing it up for me!

**Pull request merged**